README.md
# iTunesSearch - block based iTunes store communication for iOS and Mac OS X

[![Badge w/ Version](https://cocoapod-badges.herokuapp.com/v/iTunesSearch/badge.png)](http://cocoadocs.org/docsets/iTunesSearch)
[![Badge w/ Platform](https://cocoapod-badges.herokuapp.com/p/iTunesSearch/badge.svg)](http://cocoadocs.org/docsets/iTunesSearch)

A library for communicating with the iTunes store.

### Features
- Get albums for artists
- Get tracks for album
- Get apps by developer
- Search podcasts

And:
- Block based for easier usage
- No dependencies
- Result caching

## Usage
```objective-c
// Set the PHG Affiliate Token info
[ItunesSearch sharedInstance].affiliateToken = @"xxx";

// Get artist info
[[ItunesSearch sharedInstance] getAlbumsForArtist:@"Pink Floyd" limitOrNil:@20 successHandler:^(NSArray *result) {
    NSLog(@"result: %@", result);
} failureHandler:^(NSError *error) {
    NSLog(@"error: %@", error);
}];
```

See the included iOS project for examples.


## Installation
You can install iTunesSearch with [CocoaPods](http://cocoapods.org). You can also get the code and drag the iTunesSearch subfolder into your Xcode project.

### Requirements
* iTunesSearch is built using ARC and modern Objective-C syntax. You will need Xcode 4.4 or higher to use it in your project.
* iTunesSearch uses NSJSONSerialization and thus needs iOS 5 or higher.


## Issues and questions
Have a bug? Please [create an issue on GitHub](https://github.com/gangverk/iTunesSearch/issues)!


## Contributing
iTunesSearch is an open source project and your contribution is very much appreciated.

1. Check for [open issues](https://github.com/gangverk/iTunesSearch/issues) or [open a fresh issue](https://github.com/gangverk/iTunesSearch/issues/new) to start a discussion around a feature idea or a bug.
2. Fork the [repository on Github](https://github.com/gangverk/iTunesSearch) and make your changes on the **develop** branch (or branch off of it).
3. Make sure to add yourself to AUTHORS and send a pull request.


## Apps using iTunesSearch
* Last.fm Scrobbler

Are you using iTunesSearch in your iOS or Mac OS X app? Send a pull request with an updated README.md file to be included.


## License
iTunesSearch is available under the MIT license. See the LICENSE file for more info.


CHANGELOG
1.1.0 - May 12, 2015
- Added getAppsWithIds method

1.0.0 - May 12, 2015
- Added podcast search.
- Added PHG Campaign Token support.

0.1.7 - April 15, 2014
- Switched from TradeDoubler to PHG for affiliate program.

0.1.6 - March 12, 2014
- Added method for searching for apps via developer id.

0.1.5 - March 3, 2014
- Added method for searching for apps via app name and developer name.

0.1.4 - November 19, 2013
- Added possibility to supply a country code.

0.1.3 - February 26, 2013
- Moved to modern Objective-c syntax for cleaner code.
- Fixed crash if no NSLocaleCountryCode was found.

0.1.2 - February 26, 2013
- Added new method for retrieving a single track.
- Removed unnecessary validation.

0.1.1 - November 26, 2012
- Fixed caching support to ensure remote calls are not made if a cached item is found.

0.1.0 - November 26, 2012
- Added caching delegate.

0.0.3 - October 31, 2012
- Fixed getIdForArtist such that it handles a nil artist name.
- Fixed integer comparison of the limit parameter in various methods.

0.0.2 - October 24, 2012
- Added a new method for retrieving multiple albums and their tracks.

0.0.1 - October 1, 2012
- First public release on CocoaPods.


Example/AppDelegate.m
//
//  AppDelegate.m
//  Example
//
//  Created by Kevin Renskers on 28-09-12.
//  Copyright (c) 2012 Gangverk. All rights reserved.
//

#import "AppDelegate.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    // Override point for customization after application launch.
    return YES;
}
							
- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


iTunesSearch/ItunesSearch.h
//
//  ItunesSearch.h
//  iTunesSearch
//
//  Created by Piers Biddlestone on 28/09/12.
//  Copyright (c) 2012 Gangverk. All rights reserved.
//

#import <Foundation/Foundation.h>

typedef void (^ItunesSearchReturnBlockWithObject)(id result);
typedef void (^ItunesSearchReturnBlockWithArray)(NSArray *result);
typedef void (^ItunesSearchReturnBlockWithError)(NSError *error);

@protocol ItunesSearchCache <NSObject>
@optional
- (NSArray *)cachedArrayForKey:(NSString *)key;
- (void)cacheArray:(NSArray *)array forKey:(NSString *)key requestParams:(NSDictionary *)params maxAge:(NSTimeInterval)maxAge;
@end

@interface ItunesSearch : NSObject

@property (strong, nonatomic) NSString *affiliateToken;
@property (strong, nonatomic) NSString *campaignToken;
@property (strong, nonatomic) NSString *countryCode;    // default: NSLocaleCountryCode
@property (unsafe_unretained, nonatomic) id <ItunesSearchCache> cacheDelegate;
@property (nonatomic) NSTimeInterval timeoutInterval;   // default: 10
@property (nonatomic) NSTimeInterval maxCacheAge;       // default: 60*60*24

+ (ItunesSearch *)sharedInstance;
- (void)performApiCallForMethod:(NSString*)method withParams:(NSDictionary *)params andFilters:(NSDictionary *)filters successHandler:(ItunesSearchReturnBlockWithObject)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;
- (void)performApiCallForMethod:(NSString*)method useCache:(BOOL)useCache withParams:(NSDictionary *)params andFilters:(NSDictionary *)filters successHandler:(ItunesSearchReturnBlockWithObject)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;

#pragma mark - Album methods

- (void)getTracksForAlbums:(NSArray *)albumIds limitOrNil:(NSNumber *)limit sucessHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;

#pragma mark - Artist methods

- (void)getAlbumsForArtist:(NSNumber *)artistId limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;
- (void)getAlbumWithArtist:(NSString *)artistName andName:(NSString *)albumName limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;
- (void)getIdForArtist:(NSString *)artist successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;

#pragma mark - Track methods

- (void)getTrackWithName:(NSString *)trackName artist:(NSString *)artist album:(NSString *)album limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;

#pragma mark - App methods

- (void)getAppWithName:(NSString *)appName developer:(NSString*)developer limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;
- (void)getAppsByDeveloper:(NSString *)developerId limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;

#pragma mark - Podcast methods

- (void)getPodcastWithName:(NSString *)podcast artist:(NSString *)artist limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler;

@end


iTunesSearch/ItunesSearch.m
//
//  ItunesSearch.m
//  iTunesSearch
//
//  Created by Piers Biddlestone on 28/09/12.
//  Copyright (c) 2012 Gangverk. All rights reserved.
//

#include <CommonCrypto/CommonDigest.h>
#import "ItunesSearch.h"

#define API_URL @"https://itunes.apple.com/"

@interface ItunesSearch ()
@property (nonatomic, strong) NSOperationQueue *queue;
@end

@implementation ItunesSearch

#pragma mark - Initialization

+ (ItunesSearch *)sharedInstance {
    static dispatch_once_t pred;
    static ItunesSearch *sharedInstance = nil;
    dispatch_once(&pred, ^{ sharedInstance = [[self alloc] init]; });
    return sharedInstance;
}

- (id)init {
    self = [super init];
    if (self) {
        self.affiliateToken = @"";
        self.campaignToken = @"";
        self.queue = [[NSOperationQueue alloc] init];
        self.timeoutInterval = 10;
        self.maxCacheAge = (60 * 60 * 24);
    }
    return self;
}

#pragma mark - Private methods

- (NSString *)md5sumFromString:(NSString *)string {
	unsigned char digest[CC_MD5_DIGEST_LENGTH], i;
	CC_MD5([string UTF8String], (CC_LONG)[string lengthOfBytesUsingEncoding:NSUTF8StringEncoding], digest);
	NSMutableString *ms = [NSMutableString string];
	for (i=0;i<CC_MD5_DIGEST_LENGTH;i++) {
		[ms appendFormat: @"%02x", (int)(digest[i])];
	}
	return [ms copy];
}

- (NSString*)urlEscapeString:(id)unencodedString {
    if ([unencodedString isKindOfClass:[NSString class]]) {
        CFStringRef originalStringRef = (__bridge_retained CFStringRef)unencodedString;
        NSString *s = (__bridge_transfer NSString *)CFURLCreateStringByAddingPercentEscapes(NULL,originalStringRef, NULL, NULL,kCFStringEncodingUTF8);
        CFRelease(originalStringRef);
        return s;
    }
    return unencodedString;
}

- (NSString *)forceString:(NSString *)value {
    if (!value) return @"";
    return value;
}

- (NSString *)countryCode {
    if (!_countryCode) _countryCode = [[NSLocale currentLocale] objectForKey:NSLocaleCountryCode];
    return _countryCode;
}

- (void)performApiCallForMethod:(NSString*)method
                     withParams:(NSDictionary *)params
                     andFilters:(NSDictionary *)filters
                 successHandler:(ItunesSearchReturnBlockWithObject)successHandler
                 failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {

    [self performApiCallForMethod:method
                         useCache:YES
                       withParams:params
                       andFilters:filters
                   successHandler:successHandler
                   failureHandler:failureHandler];
}

- (void)performApiCallForMethod:(NSString*)method
                       useCache:(BOOL)useCache
                     withParams:(NSDictionary *)params
                     andFilters:(NSDictionary *)filters
                 successHandler:(ItunesSearchReturnBlockWithObject)successHandler
                 failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {

    NSMutableDictionary *newParams = [params mutableCopy];

    // Add affiliate token if supplied
    if (self.affiliateToken && self.affiliateToken.length > 0) {
        newParams[@"at"] = self.affiliateToken;
    }

    // Add campaign token if supplied
    if (self.campaignToken && self.campaignToken.length > 0) {
        newParams[@"ct"] = self.campaignToken;
    }
    
    // Set the user's country to get the correct price
    if (self.countryCode && self.countryCode.length) {
        newParams[@"country"] = self.countryCode;
    }

    // Convert the dict of params into an array of key=value strings
    NSMutableArray *paramsArray = [NSMutableArray arrayWithCapacity:[newParams count]];
    [newParams enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        [paramsArray addObject:[NSString stringWithFormat:@"%@=%@", [self urlEscapeString:key], [self urlEscapeString:obj]]];
    }];

    // Construct the request url
    NSString *url = [NSString stringWithFormat:@"%@%@?%@", API_URL, method, [paramsArray componentsJoinedByString:@"&"]];

    // Check if we have the object in cache
    NSString *cacheKey = [self md5sumFromString:url];
    if (useCache && self.cacheDelegate && [self.cacheDelegate respondsToSelector:@selector(cachedArrayForKey:)]) {
        NSArray *cachedArray = [self.cacheDelegate cachedArrayForKey:cacheKey];
        if (cachedArray && cachedArray.count) {
            if (successHandler) {
                successHandler(cachedArray);
            }
            return;
        }
    }

    [self _performApiCallWithURL:url
                        useCache:useCache
                       signature:cacheKey
                      withParams:newParams
                      andFilters:filters
                  successHandler:successHandler
                  failureHandler:failureHandler];
}

- (void)_performApiCallWithURL:(NSString*)url
                      useCache:(BOOL)useCache
                     signature:(NSString *)cacheKey
                    withParams:(NSDictionary *)params
                    andFilters:(NSDictionary *)filters
                successHandler:(ItunesSearchReturnBlockWithObject)successHandler
                failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {

    NSBlockOperation *op = [[NSBlockOperation alloc] init];
    [op addExecutionBlock:^{
        // Set up the http request
        NSURLRequestCachePolicy policy = NSURLRequestUseProtocolCachePolicy;
        if (!useCache) {
            policy = NSURLRequestReloadIgnoringLocalAndRemoteCacheData;
        }
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:url]
                                                               cachePolicy:policy
                                                           timeoutInterval:self.timeoutInterval];
        [request setHTTPMethod:@"GET"];

        NSHTTPURLResponse *response;
        NSError *error;
        NSData *data = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:&error];

        // Check for NSURLConnection errors
        if (error) {
            if (failureHandler) {
                [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                    failureHandler(error);
                }];
            }
            return;
        }

        // Deserialise the raw data into a JSON object
        id jsonData = [NSJSONSerialization JSONObjectWithData:data
                                                      options:0
                                                        error:&error];

        // Check for data serialization errors
        if (error) {
            if (failureHandler) {
                [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                    failureHandler(error);
                }];
            }
            return;
        }

        // Ensure a dictionary was received
        if (![jsonData isKindOfClass:[NSDictionary class]]) {
            if (failureHandler) {
                [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                    // Build an error describing the failure
                    NSMutableDictionary* details = [NSMutableDictionary dictionary];
                    [details setValue:@"Expected a dictionary as the top level object"
                               forKey:NSLocalizedDescriptionKey];
                    NSError *invalidTopLevel = [NSError errorWithDomain:@"ItunesSearch"
                                                                   code:101
                                                               userInfo:details];

                    // Execute the failure handler
                    failureHandler(invalidTopLevel);
                }];
            }

            return;
        }

        // Extract the results from the returned data
        NSArray *filteredResults = nil;
        if (jsonData && [jsonData count] > 0) {
            // Pull out the results object
            NSArray *results = jsonData[@"results"];

            // Sanity check the results
            if (results) {
                // Apply filters to the results if supplied
                if (filters && [filters count] > 0) {
                    NSMutableArray *predicates = [NSMutableArray array];
                    // Construct a case-insensitive predicate for each filter
                    for (id key in [filters allKeys]) {
                        [predicates addObject:[NSPredicate predicateWithFormat:
                                               @"%K ==[c] %@",
                                               key,
                                               [filters valueForKey:key]]];
                    }

                    // Apply the predicates
                    filteredResults = [results filteredArrayUsingPredicate:
                                       [NSCompoundPredicate andPredicateWithSubpredicates:predicates]];
                } else {
                    // No filtering necessary
                    filteredResults = results;
                }
            }
        }

        // Add to cache
        if (useCache &&
            self.cacheDelegate &&
            [self.cacheDelegate respondsToSelector:@selector(cacheArray:forKey:requestParams:maxAge:)]) {
            [self.cacheDelegate cacheArray:filteredResults forKey:cacheKey requestParams:params maxAge:self.maxCacheAge];
        }

        // Send the results to the success handler
        if (successHandler) {
            [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                successHandler(filteredResults);
            }];
        }
    }];

    [self.queue addOperation:op];
}

#pragma mark - Album methods

- (void)getTracksForAlbums:(NSArray *)albumIds limitOrNil:(NSNumber *)limit sucessHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {
    // Set up the request parameters
    NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
    params[@"id"] = [albumIds componentsJoinedByString:@","];
    params[@"entity"] = @"song";

    // Add the limit if supplied
    if (limit && limit > 0) {
        params[@"limit"] = limit;
    }

    [self performApiCallForMethod:@"lookup"
                       withParams:params
                       andFilters:nil
                   successHandler:successHandler
                   failureHandler:failureHandler];
}

#pragma mark - Artist methods

- (void)getAlbumsForArtist:(NSNumber *)artistId limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {
    // Set up the request paramters
    NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
    params[@"id"] = [self forceString:[artistId stringValue]];
    params[@"media"] = @"music";
    params[@"entity"] = @"album";

    // Set up the results filter
    NSDictionary *filters = @{
        @"wrapperType": @"collection",
        @"collectionType": @"album"
    };

    // Add the limit if supplied
    if (limit && [limit integerValue] > 0) {
        params[@"limit"] = limit;
    }

    [self performApiCallForMethod:@"lookup"
                       withParams:params
                       andFilters:filters
                   successHandler:successHandler
                   failureHandler:failureHandler];
}

- (void)getAlbumWithArtist:(NSString *)artistName andName:(NSString *)albumName limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {
    // Set up the request paramters
    NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
    params[@"term"] = [self forceString:artistName];
    params[@"media"] = @"music";
    params[@"entity"] = @"album";
    params[@"attribute"] = @"artistTerm";

    // Set up the results filter
    NSDictionary *filters = @{
        @"wrapperType": @"collection",
        @"collectionType": @"album",
        @"collectionName": albumName
    };

    // Add the limit if supplied
    if (limit && [limit integerValue] > 0) {
        params[@"limit"] = limit;
    }

    [self performApiCallForMethod:@"search"
                       withParams:params
                       andFilters:filters
                   successHandler:successHandler
                   failureHandler:failureHandler];
}

- (void)getIdForArtist:(NSString *)artist successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {
    // Set up the request paramters
    NSDictionary *params = @{
        @"term": [self forceString:artist],
        @"media": @"music",
        @"entity": @"musicArtist",
        @"attribute": @"artistTerm",
    };

    // Set up the results filter
    NSDictionary *filters = @{ @"artistName": [self forceString:artist] };

    [self performApiCallForMethod:@"search"
                       withParams:params
                       andFilters:filters
                   successHandler:successHandler
                   failureHandler:failureHandler];
}

#pragma mark - Track methods

- (void)getTrackWithName:(NSString *)trackName artist:(NSString *)artist album:(NSString *)album limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {
    // Ensure only valid objects are used in the search
    NSMutableArray *searchParameters = [NSMutableArray array];
    if (trackName) {
        [searchParameters addObject:trackName];
    }

    if (artist) {
        [searchParameters addObject:artist];
    }

    if (album) {
        [searchParameters addObject:album];
    }

    // Build the search term
    NSString *searchTerm = [searchParameters componentsJoinedByString:@"+"];

    if (searchTerm && [searchTerm length] > 0) {
        // Set up the request paramters
        NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
        params[@"term"] = [self forceString:searchTerm];
        params[@"media"] = @"music";
        params[@"entity"] = @"song";

        // Add the limit if supplied
        if (limit && limit > 0) {
            params[@"limit"] = limit;
        }

        [self performApiCallForMethod:@"search"
                           withParams:params
                           andFilters:nil
                       successHandler:successHandler
                       failureHandler:failureHandler];
    } else {
        if (successHandler) {
            return successHandler(nil);
        }
    }
}

#pragma mark - App methods

- (void)getAppWithName:(NSString *)appName developer:(NSString *)developer limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {
    // Ensure only valid objects are used in the search
    NSMutableArray *searchParameters = [NSMutableArray array];

    if (appName) {
        [searchParameters addObject:appName];
    }

    if (developer) {
        [searchParameters addObject:developer];
    }

    // Build the search term
    NSString *searchTerm = [searchParameters componentsJoinedByString:@"+"];

    if (searchTerm && [searchTerm length] > 0) {
        // Set up the request paramters
        NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
        params[@"term"] = [self forceString:searchTerm];
        params[@"media"] = @"software";

        // Add the limit if supplied
        if (limit && limit > 0) {
            params[@"limit"] = limit;
        }

        [self performApiCallForMethod:@"search"
                           withParams:params
                           andFilters:nil
                       successHandler:successHandler
                       failureHandler:failureHandler];
    } else {
        if (successHandler) {
            return successHandler(nil);
        }
    }
}

- (void)getAppsByDeveloper:(NSString *)developerId limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {
    // Ensure only valid developerId is used in the search
    if (developerId && [developerId length] > 0) {
        // Set up the request paramters
        NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
        params[@"id"] = developerId;
        params[@"entity"] = @"software";

        // Add the limit if supplied
        if (limit && limit > 0) {
            params[@"limit"] = limit;
        }

        [self performApiCallForMethod:@"lookup"
                           withParams:params
                           andFilters:nil
                       successHandler:successHandler
                       failureHandler:failureHandler];
    } else {
        if (successHandler) {
            return successHandler(nil);
        }
    }
}

- (void)getAppsWithIds:(NSArray *)appIds successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {
    // Build a string of App Ids
    NSString *appIdsString = [appIds componentsJoinedByString:@","];
    
    if (appIdsString && [appIdsString length] > 0) {
        // Set up the request paramters
        NSDictionary * params = @{
                                  @"id": [self forceString:appIdsString],
                                  @"media": @"software"
                                  };
        
        [self performApiCallForMethod:@"lookup"
                           withParams:params
                           andFilters:nil
                       successHandler:successHandler
                       failureHandler:failureHandler];
    } else {
        if (successHandler) {
            return successHandler(nil);
        }
    }
}

#pragma mark - Podcast methods

- (void)getPodcastWithName:(NSString *)podcast artist:(NSString *)artist limitOrNil:(NSNumber *)limit successHandler:(ItunesSearchReturnBlockWithArray)successHandler failureHandler:(ItunesSearchReturnBlockWithError)failureHandler {

    NSMutableArray *searchParameters = [NSMutableArray array];
    if (podcast) {
        [searchParameters addObject:podcast];
    }
    
    if (artist) {
        [searchParameters addObject:artist];
    }
    
    // Build the search term
    NSString *searchTerm = [searchParameters componentsJoinedByString:@"+"];
    
    if (searchTerm && [searchTerm length] > 0) {
        // Set up the request paramters
        NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
        params[@"term"] = [self forceString:searchTerm];
        params[@"media"] = @"podcast";
        params[@"entity"] = @"podcast";
        
        // Add the limit if supplied
        if (limit && limit > 0) {
            params[@"limit"] = limit;
        }
        
        [self performApiCallForMethod:@"search"
                           withParams:params
                           andFilters:nil
                       successHandler:successHandler
                       failureHandler:failureHandler];
    } else {
        if (successHandler) {
            return successHandler(nil);
        }
    }
}


@end
